File: packages/compiler/src/generators/event.ts
import {
  IRNodeTypes,
  type OperationNode,
  type SetDynamicEventsIRNode,
  type SetEventIRNode,
  type SimpleExpressionNode,
} from '../ir'
import {
  DELIMITERS_OBJECT_NEWLINE,
  genCall,
  genMulti,
  isFnExpression,
  isMemberExpression,
  NEWLINE,
  type CodeFragment,
} from '../utils'
import type { CodegenContext } from '../generate'
import { genExpression } from './expression'

export function genSetEvent(
  oper: SetEventIRNode,
  context: CodegenContext,
): CodeFragment[] {
  const { helper } = context
  const { element, key, keyOverride, value, modifiers, delegate, effect } = oper

  const name = genName()
  const handler = genEventHandler(context, value, modifiers)
  const eventOptions = genEventOptions()

  if (delegate) {
    // key is static
    context.delegates.add(key.content)
    // if this is the only delegated event of this name on this element,
    // we can generate optimized handler attachment code
    // e.g. n1.$evtclick = () => {}
    if (!context.block.operation.some(isSameDelegateEvent)) {
      return [NEWLINE, `n${element}.$evt${key.content} = `, ...handler]
    }
  }

  return [
    NEWLINE,
    ...genCall(
      helper(delegate ? 'delegate' : 'on'),
      `n${element}`,
      name,
      handler,
      eventOptions,
    ),
  ]

  function genName(): CodeFragment[] {
    const expr = genExpression(key, context)
    if (keyOverride) {
      // TODO unit test
      const find = JSON.stringify(keyOverride[0])
      const replacement = JSON.stringify(keyOverride[1])
      const wrapped: CodeFragment[] = ['(', ...expr, ')']
      return [...wrapped, ` === ${find} ? ${replacement} : `, ...wrapped]
    } else {
      return genExpression(key, context)
    }
  }

  function genEventOptions(): CodeFragment[] | undefined {
    const { options } = modifiers
    if (!options.length && !effect) return

    return genMulti(
      DELIMITERS_OBJECT_NEWLINE,
      effect && ['effect: true'],
      ...options.map((option): CodeFragment[] => [`${option}: true`]),
    )
  }

  function isSameDelegateEvent(op: OperationNode) {
    if (
      op.type === IRNodeTypes.SET_EVENT &&
      op !== oper &&
      op.delegate &&
      op.element === oper.element &&
      op.key.content === key.content
    ) {
      return true
    }
  }
}

export function genSetDynamicEvents(
  oper: SetDynamicEventsIRNode,
  context: CodegenContext,
): CodeFragment[] {
  const { helper } = context
  return [
    NEWLINE,
    ...genCall(
      helper('setDynamicEvents'),
      `n${oper.element}`,
      genExpression(oper.event, context),
    ),
  ]
}

export function genEventHandler(
  context: CodegenContext,
  value: SimpleExpressionNode | undefined,
  modifiers: {
    nonKeys: string[]
    keys: string[]
  } = { nonKeys: [], keys: [] },
  // passed as component prop - need additional wrap
  extraWrap: boolean = false,
): CodeFragment[] {
  let handlerExp: CodeFragment[] = [`() => {}`]
  if (value && value.content.trim()) {
    // Determine how the handler should be wrapped so it always reference the
    // latest value when invoked.
    if (isMemberExpression(value)) {
      // e.g. @click="foo.bar"
      handlerExp = genExpression(value, context)
      if (!extraWrap) {
        // non constant, wrap with invocation as `e => foo.bar(e)`
        // when passing as component handler, access is always dynamic so we
        // can skip this
        handlerExp = [`e => `, ...handlerExp, `(e)`]
      }
    } else if (isFnExpression(value)) {
      // Fn expression: @click="e => foo(e)"
      // no need to wrap in this case
      handlerExp = genExpression(value, context)
    } else {
      // inline statement
      const hasMultipleStatements = value.content.includes(`;`)
      handlerExp = [
        '() => ',
        hasMultipleStatements ? '{' : '(',
        ...genExpression(value, context),
        hasMultipleStatements ? '}' : ')',
      ]
    }
  }

  const { keys, nonKeys } = modifiers
  if (nonKeys.length)
    handlerExp = genWithModifiers(context, handlerExp, nonKeys)
  if (keys.length) handlerExp = genWithKeys(context, handlerExp, keys)

  if (extraWrap) handlerExp.unshift(`() => `)
  return handlerExp
}

function genWithModifiers(
  context: CodegenContext,
  handler: CodeFragment[],
  nonKeys: string[],
): CodeFragment[] {
  return genCall(
    context.helper('withModifiers'),
    handler,
    JSON.stringify(nonKeys),
  )
}

function genWithKeys(
  context: CodegenContext,
  handler: CodeFragment[],
  keys: string[],
): CodeFragment[] {
  return genCall(context.helper('withKeys'), handler, JSON.stringify(keys))
}

File: packages/compiler-rs/src/generate/event.rs
use oxc_ast::NONE;
use oxc_ast::ast::{
  AssignmentTarget, Expression, FormalParameterKind, ObjectPropertyKind, PropertyKind, Statement,
};
use oxc_span::SPAN;

use crate::generate::CodegenContext;
use crate::generate::expression::gen_expression;
use crate::ir::index::{Modifiers, SetDynamicEventsIRNode, SetEventIRNode, SimpleExpressionNode};

pub fn gen_set_event<'a>(
  oper: SetEventIRNode<'a>,
  context: &'a CodegenContext<'a>,
  event_opers: &Vec<SetEventIRNode>,
) -> Statement<'a> {
  let ast = &context.ast;
  let SetEventIRNode {
    element,
    key,
    value,
    modifiers: Modifiers {
      options,
      keys,
      non_keys,
    },
    delegate,
    effect,
    ..
  } = oper;

  let key_content = key.content.clone();
  let oper_key_strat = key.loc.start;
  let name = gen_expression(key, context, None, None);
  let event_options = if options.len() == 0 && !effect {
    None
  } else {
    let mut properties = ast.vec();
    if effect {
      properties.push(ObjectPropertyKind::ObjectProperty(
        ast.alloc_object_property(
          SPAN,
          PropertyKind::Init,
          ast.property_key_static_identifier(SPAN, ast.atom("effect")),
          ast.expression_boolean_literal(SPAN, true),
          false,
          false,
          false,
        ),
      ))
    }
    properties.extend(options.into_iter().map(|option| {
      ObjectPropertyKind::ObjectProperty(ast.alloc_object_property(
        SPAN,
        PropertyKind::Init,
        ast.property_key_static_identifier(SPAN, ast.atom(&option)),
        ast.expression_boolean_literal(SPAN, true),
        false,
        false,
        false,
      ))
    }));
    Some(ast.expression_object(SPAN, properties))
  };
  let handler = gen_event_handler(context, value, keys, non_keys, false);

  if delegate {
    // key is static
    context
      .options
      .delegates
      .borrow_mut()
      .insert(key_content.clone());
    // if this is the only delegated event of this name on this element,
    // we can generate optimized handler attachment code
    // e.g. n1.$evtclick = () => {}
    if !event_opers.iter().any(|op| {
      if op.key.loc.start != oper_key_strat
        && op.delegate
        && op.element == oper.element
        && op.key.content == key_content
      {
        true
      } else {
        false
      }
    }) {
      return ast.statement_expression(
        SPAN,
        ast.expression_assignment(
          SPAN,
          oxc_ast::ast::AssignmentOperator::Assign,
          AssignmentTarget::StaticMemberExpression(ast.alloc_static_member_expression(
            SPAN,
            ast.expression_identifier(SPAN, ast.atom(&format!("n{element}"))),
            ast.identifier_name(SPAN, ast.atom(&format!("$evt{key_content}"))),
            false,
          )),
          handler,
        ),
      );
    }
  }

  let mut arguments = ast.vec();
  arguments.push(
    ast
      .expression_identifier(SPAN, ast.atom(&format!("n{element}")))
      .into(),
  );
  arguments.push(name.into());
  arguments.push(handler.into());
  if let Some(event_options) = event_options {
    arguments.push(event_options.into());
  }

  ast.statement_expression(
    SPAN,
    ast.expression_call(
      SPAN,
      ast.expression_identifier(
        SPAN,
        ast.atom(&context.helper(if delegate { "delegate" } else { "on" })),
      ),
      NONE,
      arguments,
      false,
    ),
  )
}

pub fn gen_event_handler<'a>(
  context: &'a CodegenContext<'a>,
  value: Option<SimpleExpressionNode<'a>>,
  keys: Vec<String>,
  non_keys: Vec<String>,
  // passed as component prop - need additional wrap
  extra_wrap: bool,
) -> Expression<'a> {
  let ast = &context.ast;
  let mut handler_exp = if let Some(value) = value
    && !value.content.trim().is_empty()
  {
    gen_expression(value, context, None, None)
  } else {
    ast.expression_arrow_function(
      SPAN,
      false,
      false,
      NONE,
      ast.formal_parameters(
        SPAN,
        FormalParameterKind::ArrowFormalParameters,
        ast.vec(),
        NONE,
      ),
      NONE,
      ast.function_body(SPAN, ast.vec(), ast.vec()),
    )
  };

  if non_keys.len() > 0 {
    handler_exp = ast.expression_call(
      SPAN,
      ast.expression_identifier(SPAN, ast.atom(&context.helper("withModifiers"))),
      NONE,
      ast.vec_from_array([
        handler_exp.into(),
        ast
          .expression_array(
            SPAN,
            ast.vec_from_iter(non_keys.into_iter().map(|key| {
              ast
                .expression_string_literal(SPAN, ast.atom(&key), None)
                .into()
            })),
          )
          .into(),
      ]),
      false,
    )
  }

  if keys.len() > 0 {
    handler_exp = ast.expression_call(
      SPAN,
      ast.expression_identifier(SPAN, ast.atom(&context.helper("withKeys"))),
      NONE,
      ast.vec_from_array([
        handler_exp.into(),
        ast
          .expression_array(
            SPAN,
            ast.vec_from_iter(keys.into_iter().map(|key| {
              ast
                .expression_string_literal(SPAN, ast.atom(&key), None)
                .into()
            })),
          )
          .into(),
      ]),
      false,
    )
  }

  if extra_wrap {
    handler_exp = ast.expression_arrow_function(
      SPAN,
      true,
      false,
      NONE,
      ast.formal_parameters(
        SPAN,
        FormalParameterKind::ArrowFormalParameters,
        ast.vec(),
        NONE,
      ),
      NONE,
      ast.function_body(
        SPAN,
        ast.vec(),
        ast.vec1(ast.statement_expression(SPAN, handler_exp)),
      ),
    )
  }
  handler_exp
}

pub fn gen_set_dynamic_events<'a>(
  oper: SetDynamicEventsIRNode<'a>,
  context: &'a CodegenContext<'a>,
) -> Statement<'a> {
  let ast = &context.ast;
  ast.statement_expression(
    SPAN,
    ast.expression_call(
      SPAN,
      ast.expression_identifier(SPAN, ast.atom(&context.helper("setDynamicEvents"))),
      NONE,
      ast.vec_from_array([
        ast
          .expression_identifier(SPAN, ast.atom(&format!("n{}", oper.element)))
          .into(),
        gen_expression(oper.value, context, None, None).into(),
      ]),
      false,
    ),
  )
}
